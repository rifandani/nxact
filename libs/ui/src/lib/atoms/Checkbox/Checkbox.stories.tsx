import type { CheckedState } from '@radix-ui/react-checkbox';
import type { ComponentMeta, ComponentStoryFn } from '@storybook/react';
import { useState } from 'react';
import { Checkbox } from './Checkbox';

const Story: ComponentMeta<typeof Checkbox> = {
  title: 'atoms/Checkbox',
  component: Checkbox,
  argTypes: {
    asChild: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    defaultChecked: {
      control: { type: 'boolean' },
    },
    checked: {
      control: { type: 'select', options: [true, false, 'indeterminate'] },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    required: {
      control: { type: 'boolean' },
    },
    name: {
      control: { type: 'string' },
    },
    value: {
      control: { type: 'string' },
      defaultValue: 'on',
    },
  },
};
export default Story;

const Template: ComponentStoryFn<typeof Checkbox> = (args) => {
  const [checked, setChecked] = useState<CheckedState>('indeterminate');

  return (
    <div className="flex items-center space-x-2">
      <Checkbox
        id="terms"
        checked={checked}
        onCheckedChange={setChecked}
        {...args}
      />
      <label
        htmlFor="terms"
        className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
      >
        Accept terms and conditions
      </label>
    </div>
  );
};

const IndeterminateTemplate: ComponentStoryFn<typeof Checkbox> = (args) => {
  const [checked, setChecked] = useState<CheckedState>('indeterminate');

  return (
    <>
      <Checkbox checked={checked} onCheckedChange={setChecked} {...args} />

      <button
        type="button"
        onClick={() =>
          setChecked((prev) =>
            prev === 'indeterminate' ? false : 'indeterminate'
          )
        }
      >
        Toggle
      </button>
    </>
  );
};

export const Default = Template.bind({});
export const Indeterminate = IndeterminateTemplate.bind({});
