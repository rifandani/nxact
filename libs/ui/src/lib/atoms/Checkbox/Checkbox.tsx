import * as CheckboxPrimitive from '@radix-ui/react-checkbox';
import React from 'react';

import { cn } from 'libs/ui/src/utils/cn.util';
import '../../../index.css';
import { Check, Minus } from '../Icon/Icon';

const Checkbox = React.forwardRef<
  React.ElementRef<typeof CheckboxPrimitive.Root>,
  React.ComponentPropsWithoutRef<typeof CheckboxPrimitive.Root>
>(({ className, checked, ...props }, ref) => (
  <CheckboxPrimitive.Root
    ref={ref}
    className={cn(
      'peer h-4 w-4 shrink-0 rounded-sm border border-slate-300 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-700 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-900',
      className
    )}
    {...props}
  >
    <CheckboxPrimitive.Indicator
      className={cn('flex items-center justify-center')}
    >
      {checked === 'indeterminate' && <Minus className="h-4 w-4" />}
      {checked === true && <Check className="h-4 w-4" />}
    </CheckboxPrimitive.Indicator>
  </CheckboxPrimitive.Root>
));

Checkbox.displayName = CheckboxPrimitive.Root.displayName;

export { Checkbox };
