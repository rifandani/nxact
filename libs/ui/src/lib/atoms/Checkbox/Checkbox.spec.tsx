import { Checkbox } from 'libs/ui/src/lib/atoms/Checkbox/Checkbox';
import renderer from 'react-test-renderer';

describe('Checkbox component', () => {
  it('should render correctly', () => {
    const tree = renderer.create(<Checkbox />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when component is disabled', () => {
    const tree = renderer.create(<Checkbox disabled />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly when used alongside some label', () => {
    const tree = renderer
      .create(
        <section className="flex items-center space-x-2">
          <Checkbox id="terms" />
          <label
            htmlFor="terms"
            className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
          >
            Accept terms and conditions
          </label>
        </section>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
