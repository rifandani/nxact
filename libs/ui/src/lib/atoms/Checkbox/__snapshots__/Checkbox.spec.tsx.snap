// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render correctly 1`] = `
Array [
  <button
    aria-checked={false}
    className="peer h-4 w-4 shrink-0 rounded-sm border border-slate-300 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-700 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-900"
    data-state="unchecked"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  />,
  <input
    aria-hidden={true}
    defaultChecked={false}
    style={
      Object {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Checkbox component should render correctly when component is disabled 1`] = `
Array [
  <button
    aria-checked={false}
    className="peer h-4 w-4 shrink-0 rounded-sm border border-slate-300 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-700 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-900"
    data-disabled=""
    data-state="unchecked"
    disabled={true}
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  />,
  <input
    aria-hidden={true}
    defaultChecked={false}
    disabled={true}
    style={
      Object {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />,
]
`;

exports[`Checkbox component should render correctly when used alongside some label 1`] = `
<section
  className="flex items-center space-x-2"
>
  <button
    aria-checked={false}
    className="peer h-4 w-4 shrink-0 rounded-sm border border-slate-300 focus:outline-none focus:ring-2 focus:ring-slate-400 focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 dark:border-slate-700 dark:text-slate-50 dark:focus:ring-slate-400 dark:focus:ring-offset-slate-900"
    data-state="unchecked"
    id="terms"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="checkbox"
    type="button"
    value="on"
  />
  <input
    aria-hidden={true}
    defaultChecked={false}
    style={
      Object {
        "margin": 0,
        "opacity": 0,
        "pointerEvents": "none",
        "position": "absolute",
        "transform": "translateX(-100%)",
      }
    }
    tabIndex={-1}
    type="checkbox"
    value="on"
  />
  <label
    className="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
    htmlFor="terms"
  >
    Accept terms and conditions
  </label>
</section>
`;
